design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,169 @@
+   create OSDs on nodes with as much complexity as Ceph is able.
+3. Do not introduce unnecessary complexity into Rook or into Rook's CephCluster CRD.
+
+## Proposal
+
+### Removing OSDs
+Removing OSDs is the simplest place to start with this design. When removing an OSD or OSDs via the
+Rook Orchestrator Module, the Ceph mgr running in Kubernetes will run a Kubernetes Job that will run
+the required `ceph-volume` command necessary to remove the OSD(s) from a node as desired by the Ceph
+user. The Orchestrator Module will also perform whatever Ceph operations are necessary to remove the
+OSD as would be expected by the user. This will replace the manual, imperative steps currently
+required by Rook-Ceph users to remove OSDs from the storage cluster.
+
+#### OSD removal in the CephCluster resource
+The CephCluster CRD allows adding disks to nodes in several places including disks on all nodes in
+the Kubernetes cluster, to a specific node, or via disk pattern matching (in two different ways). To
+avoid the complexity of having to manage the logic for determining how to remove a node from the
+CephCluster resource, OSD removal will **not** alter the CephCluster CRD. While doing so would be
+convenient for users, there are a lot of opportunities in development to miss corner cases or
+introduce logic bugs. This may mean that Rook could re-add the disk after the disk is removed via
+the Ceph Orchestrator interface if the CephCluster CRD is configured to do so, and the user should
+be made aware of this possibility via documentation.
+
+### Adding OSDs
+When adding an OSD or OSDs via the Rook Orchestrator Module, the Ceph mgr running in Kubernetes will
+run a Kubernetes Job that will run the required `ceph-volume` command necessary to provision the
+OSD(s) on a node node as desired by the Ceph user. This will be an imperative operation and will not
+affect the CephCluster resource. Based on arguments from
+[OSD removal in the CephCluster resource](#osd-removal-in-the-cephcluster-resource), adding OSDs via
+the Ceph orchestrator interface will not add disks to the CephCluster resource so as to maintain
+internal consistency between add and remove operations.
+
+Additionally, if OSDs were to be added via the CephCluster resource, in order to maintain goal
+number 2, the CephCluster CRD would have to be modified to either:
+
+1. Add a full understanding of `ceph-volume`'s options including Drive Groups, OR
+2. Add the ability to specify an arbitrary number of fully-formed `ceph-volume` commands to run on
+   nodes to create OSDs
+
+Both of these options conflict with goal number 3, adding unnecessary complexity to the
+CephCluster CRD which would be unwieldy for users who primarily intend to manage Rook via the CRD
+themselves and not make use of the Orchestrator Module's OSD deployment and management capabilities.
+
+### Avoiding conflicts between Rook and Ceph Orchestrator Module OSD management
```

Comment:
By adding all OSDs through the CR I hope we can avoid this whole section :)

Reply:
Agreed, I believe the entrypoint of every OSD action should be part of the CephCluster CR. Otherwise, it seems like we are implementing the same goal but differently.
Perhaps the prepare job needs to reworked to be more fined grained.

Reply:
On top of that, I believe there is a scenario where the mgr interface takes over too so if we have existing resources it should be capable of assimilating them. If we don't use the CephCluster CR that's going to be challenging.
I don't think the mgr interface/dashboard and the CephCluster CR are mutually exclusive, the mgr is just an API endpoint that consumes whatever Rook-Ceph does today.